### ISSUE TITLE
`apply_filters` returns null if the filter has an expectation

### ISSUE THREAD

**Issue by stephenharris** [Brain-WP/BrainMonkey#16]
(...unless that expectation also specifies a return value)

The documentation states that `apply_filters` will return the filtered argument. This is true, unless you add an expectation to it:

```
    Filters::expectApplied('bar')->once();
    $foo = apply_filters('foo', 'Yes!');
    $bar = apply_filters('bar', 'Yes!');
    $this->assertEquals( 'Yes!', $foo );  //Passes
    $this->assertEquals( 'Yes!', $bar ); //Fails - asserting NULL equals expected 'Yes!'
```

With an expectation you **must** provide a return value. For example:

```
    Filters::expectApplied('bar')->once()->andReturnUsing( function() {
        return func_get_arg(0);
    } );
    $filtered = apply_filters('bar', 'Yes!');
    $this->assertEquals( 'Yes!', $filtered );
```

Perhaps I just misread the documentation, but I think the documentation should highlight this limitation.

I have looked into fixing this, as you can use `byDefault()` to specify default behaviour **but**:
1. To mimic `apply_filters()` you need to use `returnUsing()` to create a closure to return the filtered argument. Unfortunately you cannot over-ride that with a `returnValue()` even if using by `byDefault()`
   
   ```
   //Default behaviour
   $mock->returnUsing(function(){
     return func_get_arg(0);
   })->byDefault();
   
   //Can't be over-ridden by client code 
   $mock->returnValue( 'foo' );
   
   //Mock will **not** return 'foo'.
   ```
   
   I think this limitation is due to the fact mockery doesn't let you mix `returnValue()` and `returnUsing()`.
2. Even if we could work around (1), `byDefault()` can only be over-ridden once. So `testApplySameFilterDifferentArguments` fails.

**Comment by Giuseppe-Mazzapica** [comment]
Thanks for this.

I am aware of the behavior, and I think it is not a bug, but a feauture :laughing: 

Reason is that this behavior is also the WordPress behavior.

In fact, if no callbacks are attached to a filter, WordPress will return the original value.
If a filter callbacks is attached, and it has no return value, the original value is changed to `null`, very likely breaking something.

Example:

```php
function give_me_foo( $foo ) {
  return apply_filters( 'foo', $foo );
}

$foo = give_me_foo( 'Foo!' )  // $foo is  'Foo!': no filters attached, value is not changed

add_filter(  'foo', function( $foo ) {
   $foo = strtoupper( $foo ); // changed, but forgot to return...
});

$foo = give_me_foo( 'Foo!' )  // $foo is  null now...
```

What shown above is something that every WordPress developer experienced: forget to return in a filter callback.

So Brain Monkey force to set expectation on return to remember us to `return ` from filters ;)

Hoever **I completely agree that documentation needs to be improved to exaplain this**.

---

_Note:_ I know that sometimes in tests one wants just to test that a filters is applied, this is why I added to Brain Monkey `Filters::applied()` method. It returns hoe many times a filter was called, without altering the returned value.

```php

public function testGiveMeFooAppliesFooFilter() {
   give_me_foo();
   give_me_foo();
   $this->assertSame( 2, Brain\Monkey\filters()->applied('foo') );`
}
```

The same thing can be done with `do_action()` that is _automatically_ mocked by Brain Monkey.

**Comment by GaryJones** [comment]
As per discussion on Slack, I find the existing behaviour confusing as well.

My code was checking that a filter was `expectApplied()`, followed by checking a function was called with some args. Since the `expectApplied()` was apparently returning `null`, having both sets of checks in a single test failed, whereas having them in separate tests made them both pass.

@stephenharris's workaround of returning the first arg worked for me here.

I did also try the `$this->assertSame( 2, Brain\Monkey\filters()->applied('foo') );` approach, and that seemed to have it's own count issues which I didn't dive into to debug.

The logic about requiring a `return` seems out of place here. This code is not testing the filter callback (where a `return` would obviously be needed), but about some other code that happens to call `apply_filters()`.

Even if the behaviour can't be changed, it should definitely be documented to make this clear.

**Comment by gmazzap** [comment]
I've played around some changes for version 2 of Brain Monkey.

There I think I will keep the path that @stephenharris proposed (default to return first argument with `byDefault()`).

And yes, doing that we can replace only once, which means that:

## Case 1, no return expectation:

```
Filters::expectApplied('foo');

$yes = apply_filters('foo', 'Yes!');

self::assertEquals( 'Yes!', $yes );
```

Will pass, because _by default_ Brain Monkey will return the first argument on expectations.

## Case 2, one return expectation:

```
Filters::expectApplied('foo')->andReturn('No!');

$yes = apply_filters('foo', 'Yes!');

self::assertEquals( 'No!', $yes );
```

Will pass, because it will be possible to replace default return value.

## Case 3, more return expectations:

```
Filters::expectApplied('foo')->once()->with('No?')->andReturn('No!');
Filters::expectApplied('foo')->once()->with('Yes?')->andReturn('Yes!');
Filters::expectApplied('foo')->once()->with('Maybe?')->andReturn('Maybe!');

$no = apply_filters('foo', 'No?');
$yes = apply_filters('foo', 'Yes?');
$maybe = apply_filters('foo', 'Maybe?');

self::assertEquals( 'No!', $no );
self::assertEquals( 'Yes!', $yes );
self::assertEquals( 'Maybe!', $maybe );
```

Will pass, because it will be possible to replace default return value more than once, for different arguments.

## Case 4, more return expectations, but not for possible cases:

```
Filters::expectApplied('foo')->once()->with('No?')->andReturn('No!');
Filters::expectApplied('foo')->once()->with('Yes?')->andReturn('Yes!');

$no = apply_filters('foo', 'No?');
$yes = apply_filters('foo', 'Yes?');
$maybe = apply_filters('foo', 'Maybe?'); // there's no expectation for this

self::assertEquals( 'No!', $no );
self::assertEquals( 'Yes!', $yes );
self::assertEquals( 'Maybe?', $maybe );
```

Will **fail**. Because when same filter is applied with different arguments either no expecation of arguments is set or it will be needed that there's an expectation for _any_ kind of argument.

I think this is a good trade-off. Even because it will always possible to add a catch-all behaviour:

## Case 5, more return expectations, with a catch-all behaviour:

```
Filters::expectApplied('foo')->once()->with('No?')->andReturn('No!');
Filters::expectApplied('foo')->once()->with('Yes?')->andReturn('Yes!');
Filters::expectApplied('foo')->zeroOrMoreTimes()->withAnyArgs()->andReturnFirstArg();

$no = apply_filters('foo', 'No?');
$yes = apply_filters('foo', 'Yes?');
$maybe = apply_filters('foo', 'Maybe?');

self::assertEquals( 'No!', $no );
self::assertEquals( 'Yes!', $yes );
self::assertEquals( 'Maybe?', $maybe );
```

Will *pass* again. Note that `andReturnFirstArg()` is not a Mockery function, but a Brain Monkey helper that I plan to add in version 2.

**Comment by GaryJones** [comment]
Every single one of these cases makes perfect sense to me - I'm fully behind this logic. The above is pretty much all the documentation needed for it as well! (Docs-driven design!)

**Comment by GaryJones** [comment]
Has this all been completed in BM v2?

**Comment by gmazzap** [comment]
It should, if I did not miss anything.

**Comment by gmazzap** [comment]
@GaryJones Actually there were cases among the 5 I described in my comment from 27 April, that weren't working. They are fixed with v2.0.3 I just released.

I also updated documentation (https://github.com/Brain-WP/BrainMonkey/blob/master/docs/migrating-from-v1.md#fixed-apply_filters-default-behavior) and wrote [a complete set of unit tests to cover all the cases](https://github.com/Brain-WP/BrainMonkey/blob/master/tests/src/Api/ApplyFiltersTest.php#L114-L200).

It is not really possible to backport these changes to version 1, and version 2 got this fixed, so I call this done and I'll close the issue.

Anyone having problem, feel free to re-open it.
